cpp/src/arrow/parquet/writer.cc

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/parquet/writer.h"
+
+#include "arrow/array.h"
+#include "arrow/types/primitive.h"
+#include "arrow/parquet/utils.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+
+namespace parquet {
+
+class FileWriter::Impl {
+ public:
+  Impl(MemoryPool* pool, std::unique_ptr<::parquet::ParquetFileWriter> writer);
+
+  Status NewRowGroup(int64_t chunk_size);
+  template <typename ParquetType, typename CType>
+  Status TypedWriteBatch(::parquet::ColumnWriter* writer, const PrimitiveArray* data);
+  Status WriteFlatColumnChunk(const PrimitiveArray* data);
+  Status Close();
+
+  virtual ~Impl() {}
+
+ private:
+  MemoryPool* pool_;
+  PoolBuffer data_buffer_;
+  PoolBuffer def_levels_buffer_;
+  std::unique_ptr<::parquet::ParquetFileWriter> writer_;
+  ::parquet::RowGroupWriter* row_group_writer_;
+};
+
+FileWriter::Impl::Impl(
+    MemoryPool* pool, std::unique_ptr<::parquet::ParquetFileWriter> writer)
+    : pool_(pool),
+      data_buffer_(pool),
+      writer_(std::move(writer)),
+      row_group_writer_(nullptr) {}
+
+Status FileWriter::Impl::NewRowGroup(int64_t chunk_size) {
+  if (row_group_writer_ != nullptr) { PARQUET_CATCH_NOT_OK(row_group_writer_->Close()); }
+  PARQUET_CATCH_NOT_OK(row_group_writer_ = writer_->AppendRowGroup(chunk_size));
+  return Status::OK();
+}
+
+template <typename ParquetType, typename CType>
+Status FileWriter::Impl::TypedWriteBatch(
+    ::parquet::ColumnWriter* column_writer, const PrimitiveArray* data) {
+  auto data_ptr = reinterpret_cast<const CType*>(data->data()->data());
+  auto writer =
+      reinterpret_cast<::parquet::TypedColumnWriter<ParquetType>*>(column_writer);
+  if (writer->descr()->max_definition_level() == 0) {
+    // no nulls, just dump the data
+    writer->WriteBatch(data->length(), nullptr, nullptr, data_ptr);
+  } else if (writer->descr()->max_definition_level() == 1) {
+    def_levels_buffer_.Resize(data->length() * sizeof(int16_t));
+    int16_t* def_levels_ptr =
+        reinterpret_cast<int16_t*>(def_levels_buffer_.mutable_data());
+    std::fill(def_levels_ptr, def_levels_ptr + data->length(), 1);
```

Comment:
It may be slightly pedantic (though conceivably slightly better performance), but you could call `std::fill` only when `null_count == 0` and set 1 or 0 below (rather than only setting 0's)